File: Point_to_Point/exercises/Python/ex1.py
from mpi4py import MPI
import sys
import numpy
import random
import os


###########################
# Initialize Communication
cw = MPI.COMM_WORLD
rank = cw.rank
size = cw.size
proc = MPI.Get_processor_name()

#############################################
# Read problem parameters from command line
ofile='tmp.dat'
n = numpy.zeros(1,dtype='i')
if (rank == 0):
    n[0] = int(sys.argv[2])
    #ofile = 'Numbers_N'+str(n[0])
    #try:
    #    os.remove(ofile)
    #except:
    #    pass


cw.Bcast(n,root=0)
ofile = 'Numbers_N'+str(n[0])


#######################
# Initialize data (random number sequence)
my_numbers=numpy.zeros(n[0],dtype='d')
for i in range(n[0]):
    my_numbers[i] = random.random()



############################################################
# Communicate, perform calculation on rank 0, and record results

if (rank == 0):
    # Open the file
    #'The value is {:0,.2f}'.format(x)

    myfile = open(ofile,"w")
    for i in range(n[0]):
        istr = "{:8d}".format(i)
        fstr = "{:.5e}".format(my_numbers[i])
        myfile.write(istr+' '+fstr+"\n")
    myfile.close()

    # Receive data
    your_numbers = numpy.zeros(n[0],dtype='d')
    cw.Recv(your_numbers, source=1,tag=1)

    # Peform calculation
    dprod = 0.0
    for i in range(n[0]):
        dprod += your_numbers[i]*my_numbers[i]

    print('The dot product is: ', dprod)

elif (rank == 1):
    cw.Send(my_numbers,dest=0,tag=1)
    fcheck = os.path.exists(ofile)
    if (fcheck):
        myfile = open(ofile,"a")
        for i in range(n[0]):
            istr = "{:8d}".format(i+n[0])
            fstr = "{:.5e}".format(my_numbers[i])
            myfile.write(istr+' '+fstr+"\n")
        myfile.close()
    else:
        print("ERROR:   Rank 1 unable to find "+ofile)

MPI.Finalize()

File: Point_to_Point/exercises/R/ex1.R
suppressMessages(library(pbdMPI, quietly = TRUE))

init()
.comm.size <- comm.size()
.comm.rank <- comm.rank()

ofile =  'tmp.dat'

if (.comm.rank == 0) {
        print(sprintf("%d MPI processes are active", .comm.size))
        args = commandArgs(TRUE)
        n = as.integer(args[1])
        print(sprintf("n is: %d", n))
        nb = bcast(n)
} else {
    n = 0
    nb = bcast(n)
}

myNumbers = runif(nb)

if (.comm.rank == 0) {
    barrier()
    yourNumbers = numeric(nb)
    recv(yourNumbers, rank.source = 1, tag = 1)
    dodProduct = myNumbers %*% yourNumbers
       
} else if (.comm.rank == 1) {
    barrier()
    send(myNumbers, rank.dest = 0, tag =1)
}

finalize()

